include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-    const bool transmitted, const float3 V, const float3 L,
-    const float3 N, const float NdotL, const float VdotL,
-    const float orientedEta, const float rcpOrientedEta, out float3 H
-  )
-  {
-    // TODO: can we optimize?
-    H = computeMicrofacetNormal(transmitted,V,L,orientedEta);
-    retval.NdotH = dot(N,H);
-    
-    // not coming from the medium (reflected) OR
-    // exiting at the macro scale AND ( (not L outside the cone of possible directions given IoR with constraint VdotH*LdotH<0.0) OR (microfacet not facing toward the macrosurface, i.e. non heightfield profile of microsurface) ) 
-    const bool valid = !transmitted || (VdotL<=-min(orientedEta,rcpOrientedEta) && _cache.NdotH>nbl::hlsl::numeric_limits::min());
-    if (valid)
-    {
-      // TODO: can we optimize?
-      retval.VdotH = dot(V,H);
-      retval.LdotH = dot(L,H);
-      retval.NdotH2 = retval.NdotH*retval.NdotH;
-      return true;
-    }
-    return false;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static bool compute(
-    out IsotropicMicrofacetCache retval,
-    const surface_interactions::Isotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta, out float3 H
-  )
-  {
-    const float NdotV = interaction.NdotV;
-    const float NdotL = _sample.NdotL;
-    const bool transmitted = nbl_glsl_isTransmissionPath(NdotV,NdotL);
-    
-    float orientedEta, rcpOrientedEta;
-    const bool backside = nbl_glsl_getOrientedEtas(orientedEta,rcpOrientedEta,NdotV,eta);
-
-    const vec3 V = interaction.V.getDirection();
-    const vec3 L = _sample.L;
-    const float VdotL = dot(V,L);
-    return nbl_glsl_calcIsotropicMicrofacetCache(_cache,transmitted,V,L,interaction.N,NdotL,VdotL,orientedEta,rcpOrientedEta,H);
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static bool compute(
-    out IsotropicMicrofacetCache retval,
-    const surface_interactions::Isotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta
-  )
-  {
-    float3 dummy;
-    return nbl_glsl_calcIsotropicMicrofacetCache(_cache,transmitted,V,L,interaction.N,NdotL,VdotL,orientedEta,rcpOrientedEta,dummy);
-  }
-
-  bool isValidVNDFMicrofacet(const bool is_bsdf, const bool transmission, const float VdotL, const float eta, const float rcp_eta)
-  {
-    return NdotH >= 0.0 && !(is_bsdf && transmission && (VdotL > -min(eta,rcp_eta)));
-  }
-
-  float VdotH;
-  float LdotH;
-  float NdotH;
-  float NdotH2;
+    using this_t = SLightSample<RayDirInfo>;
+    using ray_dir_info_type = RayDirInfo;
+    using scalar_type = typename RayDirInfo::scalar_type;
+    using vector3_type = typename RayDirInfo::vector3_type;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+
+    using isotropic_type = surface_interactions::SIsotropic<RayDirInfo>;
+    using anisotropic_type = surface_interactions::SAnisotropic<RayDirInfo>;
```

Comment:
you don't have members of this type, and the constructors you have from them should be templated to allow any concept fitting type to be used https://github.com/Devsh-Graphics-Programming/Nabla/pull/811/files#r2039076352

See
https://github.com/Devsh-Graphics-Programming/Nabla/pull/811/files#r2039552765